File: client/consumer/android/lfc/app/src/main/java/com/lfc/consumer/ui/auth/AuthViewModel.kt
package com.lfc.consumer.ui.auth

import android.content.Context
import android.net.Uri
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.lfc.consumer.data.ApiClient
import com.lfc.consumer.data.MediaUploadHelper
import com.lfc.consumer.data.local.TokenManager
import com.lfc.consumer.data.model.LoginRequest
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.MultipartBody
import okhttp3.RequestBody.Companion.asRequestBody
import okhttp3.RequestBody.Companion.toRequestBody

data class AuthUiState(
    val isLoading: Boolean = false,
    val error: String? = null,
    val isSuccess: Boolean = false,
)

class AuthViewModel(
    private val tokenManager: TokenManager,
    private val context: Context,
) : ViewModel() {
    private val api = ApiClient.createApiService(tokenManager)

    private val _uiState = MutableStateFlow(AuthUiState())
    val uiState: StateFlow<AuthUiState> = _uiState.asStateFlow()

    fun login(email: String, password: String) {
        viewModelScope.launch {
            _uiState.value = AuthUiState(isLoading = true)
            try {
                val response = api.login(LoginRequest(email, password))
                tokenManager.saveSession(
                    accessToken = response.accessToken,
                    refreshToken = response.refreshToken,
                    userId = response.user.id,
                    email = response.user.email,
                    studentId = response.user.studentId,
                )
                _uiState.value = AuthUiState(isSuccess = true)
            } catch (e: Exception) {
                _uiState.value = AuthUiState(error = parseErrorMessage(e, "登录失败"))
            }
        }
    }

    fun register(
        email: String,
        password: String,
        studentId: String,
        studentCardUri: Uri?,
    ) {
        if (studentCardUri == null) {
            _uiState.value = AuthUiState(error = "请上传学生证照片")
            return
        }

        viewModelScope.launch {
            _uiState.value = AuthUiState(isLoading = true)
            try {
                val uploadFile = MediaUploadHelper.uriToUploadFile(
                    context = context,
                    uri = studentCardUri,
                    prefix = "student_card",
                )
                val requestFile = uploadFile.file.asRequestBody(uploadFile.mimeType.toMediaTypeOrNull())
                val part = MultipartBody.Part.createFormData(
                    "studentCard",
                    uploadFile.fileName,
                    requestFile,
                )

                val response = api.register(
                    email = email.toRequestBody("text/plain".toMediaTypeOrNull()),
                    password = password.toRequestBody("text/plain".toMediaTypeOrNull()),
                    studentId = studentId.toRequestBody("text/plain".toMediaTypeOrNull()),
                    studentCard = part,
                )
                tokenManager.saveSession(
                    accessToken = response.accessToken,
                    refreshToken = response.refreshToken,
                    userId = response.user.id,
                    email = response.user.email,
                    studentId = response.user.studentId,
                )
                _uiState.value = AuthUiState(isSuccess = true)
            } catch (e: Exception) {
                _uiState.value = AuthUiState(error = parseErrorMessage(e, "注册失败"))
            }
        }
    }

    fun clearError() {
        _uiState.value = _uiState.value.copy(error = null)
    }

    private fun parseErrorMessage(e: Exception, fallback: String): String {
        if (e is retrofit2.HttpException) {
            val body = e.response()?.errorBody()?.string()
            if (!body.isNullOrBlank()) {
                runCatching {
                    com.google.gson.JsonParser.parseString(body)
                        .asJsonObject["message"]?.asString
                }.getOrNull()?.let { return it }
            }
        }
        return e.message ?: fallback
    }
}

class AuthViewModelFactory(
    private val tokenManager: TokenManager,
    private val context: Context,
) : ViewModelProvider.Factory {
    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        return AuthViewModel(tokenManager, context) as T
    }
}

File: client/consumer/ios/lfc/lfc/Features/Auth/AuthViewModel.swift
import Foundation

struct AuthUiState: Equatable {
    var isLoading = false
    var error: String?
    var isSuccess = false
}

@MainActor
final class AuthViewModel: ObservableObject {
    @Published private(set) var uiState = AuthUiState()

    private let session: SessionManager

    init(session: SessionManager) {
        self.session = session
    }

    func login(email: String, password: String) async {
        uiState = AuthUiState(isLoading: true)
        do {
            try await session.login(email: email, password: password)
            uiState = AuthUiState(isSuccess: true)
        } catch {
            uiState = AuthUiState(error: session.parseErrorMessage(from: error, fallback: "登录失败"))
        }
    }

    func register(
        email: String,
        password: String,
        realName: String,
        studentId: String,
        studentCardData: Data?,
        studentCardFileName: String?,
        studentCardMimeType: String?
    ) async {
        if realName.count < 2 {
            uiState = AuthUiState(error: "请填写真实姓名")
            return
        }
        guard let studentCardData, let studentCardFileName, let studentCardMimeType else {
            uiState = AuthUiState(error: "请上传学生证照片")
            return
        }

        uiState = AuthUiState(isLoading: true)
        do {
            try await session.register(
                email: email,
                password: password,
                realName: realName,
                studentId: studentId,
                studentCardData: studentCardData,
                studentCardFilename: studentCardFileName,
                studentCardMimeType: studentCardMimeType
            )
            uiState = AuthUiState(isSuccess: true)
        } catch {
            uiState = AuthUiState(error: session.parseErrorMessage(from: error, fallback: "注册失败"))
        }
    }

    func clearError() {
        uiState.error = nil
    }

    func resetSuccess() {
        uiState.isSuccess = false
    }
}
